application/models/organization.go

```diff
@@ -76,28 +62,98 @@ func (o *Organization) GetAuthProvider() (auth.Provider, error) {
 	return &auth.EmptyProvider{}, fmt.Errorf("unsupported auth provider type: %s", o.AuthType)
 }
 
-func FindOrgByUUID(uuid string) (Organization, error) {
+func (o *Organization) FindByUUID(uuid string) error {
 
 	if uuid == "" {
-		return Organization{}, fmt.Errorf("error: access token must not be blank")
+		return fmt.Errorf("error: access token must not be blank")
 	}
 
-	org := Organization{}
+	if err := DB.Where("uuid = ?", uuid).First(o); err != nil {
+		return fmt.Errorf("error finding org by uuid: %s", err.Error())
+	}
 
-	queryString := fmt.Sprintf("uuid = '%s'", uuid)
+	return nil
+}
 
-	if err := DB.Where(queryString).First(&org); err != nil {
-		return Organization{}, fmt.Errorf("error finding org by uuid: %s", err.Error())
+func (o *Organization) FindByDomain(domain string) error {
+	var orgDomain OrganizationDomain
+	if err := DB.Where("domain = ?", domain).First(&orgDomain); err != nil {
+		return fmt.Errorf("error finding organization_domain by domain: %s", err.Error())
 	}
 
-	return org, nil
+	if err := DB.Eager().Where("id = ?", orgDomain.OrganizationID).First(o); err != nil {
+		return fmt.Errorf("error finding organization by domain: %s", err.Error())
+	}
+
+	return nil
 }
 
-func OrganizationFindByDomain(domain string) (Organization, error) {
+func (o *Organization) AddDomain(domain string) error {
+	// make sure domain is not registered to another org first
 	var orgDomain OrganizationDomain
-	if err := DB.Eager().Where("domain = ?", domain).First(&orgDomain); err != nil {
-		return Organization{}, fmt.Errorf("error finding organization by domain: %s", err.Error())
+
+	count, err := DB.Where("domain = ?", domain).Count(&orgDomain)
+	if err != nil {
+		return err
+	}
+
+	if count > 0 {
+		return fmt.Errorf("this domain (%s) is already in use", domain)
 	}
 
-	return orgDomain.Organization, nil
+	orgDomain.Domain = domain
+	orgDomain.OrganizationID = o.ID
+	err = DB.Save(&orgDomain)
+	if err != nil {
+		return err
+	}
+
+	err = DB.Load(o, "OrganizationDomains")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *Organization) RemoveDomain(domain string) error {
+	// make sure domain belongs to org for removal
+	for _, od := range o.OrganizationDomains {
```

Comment:
I've learned not to assume that relationships are pre-loaded. Should we agree to always load all relationships, so this is a safe assumption?

Reply:
I changed it to do a direct query by org id and domain name instead of depend on relation being loaded